modules/twinkleblock.js

```diff
@@ -0,0 +1,870 @@
+        tooltip: 'An article can be linked within the notice, perhaps because it was a revert to said article that dispatched this notice. Leave empty for no article to be linked.'
+      } );
+    field_template_options.append( {
+      type: $form.find('[name=actiontype][value=block]').is(':checked') ? 'hidden' : 'input',
+      name: 'block_timer',
+      label: 'Period of blocking: ',
+      value: $form.find('[name=expiry]').val() || '',
+      tooltip: 'The period the blocking is due for, for example 24 hours, 2 weeks, indefinite etc...'
+    } );
+    field_template_options.append( {
+      type: 'input',
+      name: 'block_reason',
+      label: '"You have been blocked for ..." ',
+      tooltip: 'An optional reason, to replace the default generic reason. Only available for the generic block templates.'
+    } );
+    var $previewlink = $( '<a id="twinkleblock-preivew-link">Preview</a>' );
+    $previewlink.off("click").on("click", function(){
+      Twinkle.warn.callbacks.preview($form[0]);
+    });
+    $previewlink.css({cursor: "pointer"});
+    field_template_options.append( { type: 'div', id: 'blockpreview', label: [ $previewlink[0] ] } );
+    field_template_options.append( { type: 'div', id: 'twinkleblock-previewbox', style: 'display: none' } );
+  }
+
+  var oldfield;
+  if (field_preset) {
+    oldfield = $form.find('fieldset[name="field_preset"]')[0];
+    oldfield.parentNode.replaceChild(field_preset.render(), oldfield);
+  } else {
+    $form.find('fieldset[name="field_preset"]').hide();
+  }
+  if (field_block_options) {
+    oldfield = $form.find('fieldset[name="field_block_options"]')[0];
+    oldfield.parentNode.replaceChild(field_block_options.render(), oldfield);
+  } else {
+    $form.find('fieldset[name="field_block_options"]').hide();
+  }
+  if (field_template_options) {
+    oldfield = $form.find('fieldset[name="field_template_options"]')[0];
+    oldfield.parentNode.replaceChild(field_template_options.render(), oldfield);
+    result.previewer = new Morebits.wiki.preview($(result).find('#twinkleblock-previewbox').last()[0]);
+  } else {
+    $form.find('fieldset[name="field_template_options"]').hide();
+  }
+
+  if (Twinkle.block.currentBlockInfo) {
+    Morebits.status.init($('div[name="currentblock"] span').last()[0]);
+    Morebits.status.warn(relevantUserName + ' is already blocked', 'Submit query to reblock with supplied options');
+    Twinkle.block.callback.update_form(e, Twinkle.block.currentBlockInfo);
+  }
+};
+
+/* NOTE: warn module references these, some properties are only used by
+ *    that module, others only used by block module
+ * Keep alphabetized by key name, Twinkle.block.blockGroups establishes
+ *    the order they will appear in the interface
+ *
+ * WARNING: 'anonblock' and 'allowusertalk' are enabled by default.
+ *    To disbale, set 'hardblock' and 'disabletalk', respectively
+ */
+Twinkle.block.blockPresetsInfo = {
```

Comment:
This needs a comment documenting ALL the available keys, and what values they can take. It's true that some parts of the warn and protect modules are lacking such a comment, but that is because the maintainers (over the past few years, mainly me) know what's available. On the other hand, this is something new and complex. See the top of twinkleconfig.js for an example.